src/huggingface_hub/serialization/_torch.py

```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+        weights_only (`bool`, *optional*, defaults to `False`):
+            If True, only loads the model weights without optimizer states and other metadata.
+            Only supported in PyTorch >= 1.13.
+        map_location (`str` or `torch.device`, *optional*):
+            A `torch.device` object, string or a dict specifying how to remap storage locations. It
+            indicates the location where all tensors should be loaded.
+
+    Returns:
+        `NamedTuple`: A named tuple with `missing_keys` and `unexpected_keys` fields.
+            - `missing_keys` is a list of str containing the missing keys, i.e. keys that are in the model but not in the checkpoint.
+            - `unexpected_keys` is a list of str containing the unexpected keys, i.e. keys that are in the checkpoint but not in the model.
+
+    Raises:
+        ValueError: If the checkpoint path is invalid or if the checkpoint format cannot be determined.
+        FileNotFoundError: If the checkpoint file or directory does not exist.
+        ImportError: If required dependencies are not installed.
+
+    Example:
+    ```python
+    >>> from huggingface_hub import load_torch_model
+    >>> model = ... # A PyTorch model
+    >>> load_torch_model(model, "path/to/checkpoint")
+    ```
+    """
+    checkpoint_path = Path(checkpoint_path)
+
+    if checkpoint_path.is_file():
+        state_dict = load_state_dict_from_file(
+            checkpoint_file=checkpoint_path,
+            map_location=map_location,
+            weights_only=weights_only,
+        )
+        return model.load_state_dict(state_dict, strict=strict)
+
+    elif checkpoint_path.is_dir():
+        # Check for index file which indicates sharded checkpoint
+        filename_pattern = (
+            constants.SAFETENSORS_WEIGHTS_FILE_PATTERN if safe else constants.PYTORCH_WEIGHTS_FILE_PATTERN
+        )
+        index_file = checkpoint_path / (filename_pattern.format(suffix="") + ".index.json")
+        if index_file.is_file():
+            return load_sharded_checkpoint(
+                model=model,
+                save_directory=checkpoint_path,
+                strict=strict,
+                safe=safe,
+                weights_only=weights_only,
+            )
+
+        # Look for single model file
+        model_files = list(checkpoint_path.glob("*.safetensors" if safe else "*.bin"))
+        if len(model_files) == 1:
+            state_dict = load_state_dict_from_file(
+                checkpoint_file=model_files[0],
+                map_location=map_location,
+                weights_only=weights_only,
+            )
+            return model.load_state_dict(state_dict, strict=strict)
+
+        raise ValueError(
+            f"Directory '{checkpoint_path}' does not contain a valid checkpoint. "
+            "Expected either a sharded checkpoint with an index file, or a single model file."
+        )
+
+    else:
+        raise ValueError(f"Invalid checkpoint path: {checkpoint_path}")
```

Comment:
I feel that we could get rid of `elif checkpoint_path.is_dir():` and these 2 lines. In practice we can assume that either `checkpoint_path`.is_file()` or `checkpoint_path.is_dir()` is True. Worse case, we can add a `if not checkpoint_path.exists(): raise ...` at the beginning of the method.

Removing the elif/else would deindent a large chunk of code which is better looking IMO